mozetl/maudau/maudau.py

```diff
@@ -0,0 +1,221 @@
+from moztelemetry.standards import (
+    filter_date_range,
+    count_distinct_clientids
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+DEVELOPMENT = False
+
+STORAGE_BUCKET = "net-mozaws-prod-us-west-2-pipeline-analysis"
+STORAGE_SUB_DIR = 'spenrose/maudau' if DEVELOPMENT else 'mreid/maudau'
+MAUDAU_ROLLUP_BASENAME = "engagement_ratio.csv"
+MAUDAU_SNAPSHOT_TEMPLATE = "engagement_ratio.{}.csv"
+DASHBOARD_BUCKET = "net-mozaws-prod-metrics-data"
+DASHBOARD_KEY = "firefox-dashboard/{}".format(MAUDAU_ROLLUP_BASENAME)
+
+ACTIVITY_SUBMISSION_LAG = DT.timedelta(10)
+# MONTH as in "monthly active"
+MONTH = 28
+# Recency in days of data which should be re-counted.
+UNSTABLE_WINDOW = 30
+
+
+def format_as_submission_date(date):
+    return DT.date.strftime(date, "%Y%m%d")
+
+
+def format_spark_path(bucket, prefix):
+    return "s3://{}/{}".format(bucket, prefix)
+
+
+def get_rollup_s3_paths(basename):
+    return (STORAGE_BUCKET, os.path.join(STORAGE_SUB_DIR, basename))
+
+
+def generate_filter_parameters(end_date, days_back):
+    d = {}
+    min_activity_date = end_date - DT.timedelta(days_back)
+    d['min_activity_iso'] = min_activity_date.isoformat()
+    d['max_activity_iso'] = (end_date + DT.timedelta(1)).isoformat()
+
+    d['min_submission_string'] = format_as_submission_date(min_activity_date)
+    max_submission_date = end_date + ACTIVITY_SUBMISSION_LAG
+    d['max_submission_string'] = format_as_submission_date(max_submission_date)
+    return d
+
+
+def count_active_users(frame, end_date, days_back):
+    """
+    Tally active users according to externally-defined heuristics,
+    specifically client-reported subsession_start_date.
+
+    :frame DataFrame(): conforming to main_summary schema
+    :end_date DT.date(): the last day on which to count activity.
+    :days_back int: how far back to go; 0 for daily and 28 for monthly.
+
+    Note that return values are not stable over time as activity trickles
+    in with long submission latencies.
+
+    See https://bugzilla.mozilla.org/show_bug.cgi?id=1240849
+    """
+    params = generate_filter_parameters(end_date, days_back)
+
+    filtered = filter_date_range(
+        frame,
+        frame.subsession_start_date,
+        params['min_activity_iso'],
+        params['max_activity_iso'],
+        frame.submission_date_s3,
+        params['min_submission_string'],
+        params['max_submission_string'])
+    return count_distinct_clientids(filtered)
+
+
+def parse_last_rollup(basename, start_date=None):
+    """
+    Extract the valid old counts from basename and return them with the
+    first date that needs to be re-counted.
+    """
+    start_date = start_date or DT.date.today()
+    since = (start_date - DT.timedelta(UNSTABLE_WINDOW))
+    since = format_as_submission_date(since)
+    carryover = []
+    with open(basename) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            if row['day'] >= since:
```

Comment:
There is one more check that needs to happen for a given row - we need to make sure that the difference between `row['day']` and `row['generated']` is at least `ACTIVITY_SUBMISSION_LAG` days. That is, we want to ensure that the row was computed after the data became stable.

If we have an entry like:
`day: 20170610, dau: 100, mau: 200, generated_on: 20170615`
We should still re-compute it, even though the `day` is more than ten days ago (since it was computed with 5 days of data latency rather than the desired 10 days).

I think it would be better to split this function up into stages:
- parse the rollup to return all rows (this function minus the `break` and `since` pieces)
- filter rows to remove any "unstable" results, those being cases where `day` is recent or (`generated_on` - `day` is too low)
- Identify any resulting gaps. The previous code used `UNSTABLE_WINDOW` as an upper bound to how far back to look for such gaps, so you could do that as well, or simply process all identified gaps.

Then, we can pass in the specific set of days to re-compute to the `generate_counts` function.

These split-up functions should be fairly easy to test as well.